src/parcsr_ls/par_mgr_setup.c

```diff
@@ -1351,23 +1349,40 @@ hypre_MGRSetup( void               *mgr_vdata,
             {
                restrict_type[lev] = 2;
             }
-            hypre_MGRBuildRestrict(A_array[lev], A_FF, A_FC, CF_marker,
+            hypre_MGRBuildRestrict(A_array[lev], A_FF, A_FC, A_CF, CF_marker_array[lev],
                                    coarse_pnts_global, trunc_factor, P_max_elmts[lev],
                                    strong_threshold, max_row_sum, block_jacobi_bsize,
-                                   &RT, restrict_type[lev]);
+                                   restrict_type[lev], &Wr, &R, &RT);
+            R_array[lev]  = R;
             RT_array[lev] = RT;
             hypre_GpuProfilingPopRange();
             HYPRE_ANNOTATE_REGION_END("%s", region_name);
 
             hypre_sprintf(region_name, "RAP");
             hypre_GpuProfilingPushRange(region_name);
             HYPRE_ANNOTATE_REGION_BEGIN("%s", region_name);
-            RAP_ptr = hypre_ParCSRMatrixRAPKT(RT, A_array[lev], P, 1);
+            if (RT)
+            {
+               RAP_ptr = hypre_ParCSRMatrixRAPKT(RT, A_array[lev], P, 1);
+            }
+            else if (R)
+            {
+               AP      = hypre_ParCSRMatMat(A_array[lev], P);
```

Comment:
I believe the triple matrix product is faster (not sure if only on GPU). Could this be made to use that?

Reply:
Maybe yes. @liruipeng which is preferable? Two MatMats or a single call to `hypre_ParCSRMatrixRAPKT`?

Reply:
It depends. It may be useful to leave an option for this.

Reply:
Good idea, when is it better to use one or the other?

What should be the default?

Reply:
can you use the AMG parameter, rap2, here?

Reply:
We would need to implement it for MGR as we don't have. But yes, it's possible

Reply:
Yes, that's what I'm thinking

when is it better to use one or the other?

What should be the default?

Reply:
rap2 = 0 is the default for AMG.

Reply:
You said it depends above. In which situation would it be better to do two MatMats (rap2 = 1) instead of the the triple-mat (rap = 0)?

Reply:
It is hard to say which situation. I don't have an exact answer for this question. We have rap = 0 as default for AMG based on results on Laplacians.